File: generated-code/scala/codecraft/TcpReadWrite/src/main/scala/trans_gen_test/model/debug_interface/PrimitiveType.scala
package trans_gen_test.model.debug_interface

import trans_gen_test.util.StreamUtil

/**
 * Primitive type for debug rendering
 */
sealed abstract class PrimitiveType (val tag: Int) {
    /**
     * Write PrimitiveType to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, tag)
    }
}

object PrimitiveType {
    /**
     * Lines, number of vertices should be divisible by 2
     */
    case object LINES extends PrimitiveType(0)
    /**
     * Triangles, number of vertices should be divisible by 3
     */
    case object TRIANGLES extends PrimitiveType(1)

    /**
     * Read PrimitiveType from input stream
     */
    def readFrom(stream: java.io.InputStream): PrimitiveType =
        StreamUtil.readInt(stream) match {
            case LINES.tag => LINES
            case TRIANGLES.tag => TRIANGLES
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
}

File: generated-code/typescript/codecraft/TcpReadWrite/src/model/debug-interface/primitive-type.ts
import { Stream } from "../../stream";

/**
 * Primitive type for debug rendering
 */
export class PrimitiveType {
    readonly name: string;
    readonly tag: number;

    constructor(name: string, tag: number) {
        this.name = name;
        this.tag = tag;
    }

    /**
     * Lines, number of vertices should be divisible by 2
     */
    static readonly LINES = new PrimitiveType("LINES", 0);
    /**
     * Triangles, number of vertices should be divisible by 3
     */
    static readonly TRIANGLES = new PrimitiveType("TRIANGLES", 1);

    /**
     * Read PrimitiveType from input stream
     */
    static async readFrom(stream: Stream): Promise<PrimitiveType> {
        const tag = await stream.readInt();
        if (tag == PrimitiveType.LINES.tag) {
            return PrimitiveType.LINES;
        }
        if (tag == PrimitiveType.TRIANGLES.tag) {
            return PrimitiveType.TRIANGLES;
        }
        throw new Error("Unexpected tag value");
    }

    /**
     * Write PrimitiveType to output stream
     */
    async writeTo(stream: Stream) {
        await stream.writeInt(this.tag);
    }

    [Symbol.for('nodejs.util.inspect.custom')]() {
        return this.name;
    }
}